### ISSUE TITLE
Caused by: com.rabbitmq.client.ShutdownSignalException: channel error

### ISSUE THREAD

**Issue by durgadeep** [sstone/amqp-client#29]
Folks,

   we use this amqp-client, and the following error is filling up our logs. 

Question - Does this have a potential for memory leak ?

_DD

ERROR com.github.sstone.amqp.Consumer - null
java.io.IOException: null
    at com.rabbitmq.client.impl.AMQChannel.wrap(AMQChannel.java:106) ~[amqp-client-3.0.1.jar:na]
    at com.rabbitmq.client.impl.AMQChannel.wrap(AMQChannel.java:102) ~[amqp-client-3.0.1.jar:na]
    at com.rabbitmq.client.impl.AMQChannel.exnWrappingRpc(AMQChannel.java:124) ~[amqp-client-3.0.1.jar:na]
    at com.rabbitmq.client.impl.ChannelN.exchangeDeclarePassive(ChannelN.java:693) ~[amqp-client-3.0.1.jar:na]
    at com.rabbitmq.client.impl.ChannelN.exchangeDeclarePassive(ChannelN.java:61) ~[amqp-client-3.0.1.jar:na]
    at com.github.sstone.amqp.Amqp$.declareExchange(Amqp.scala:55) ~[amqp-client_2.10-1.1.jar:na]
    at com.github.sstone.amqp.Consumer.com$github$sstone$amqp$Consumer$$setupBinding(Consumer.scala:32) ~[amqp-client_2.10-1.1.jar:na]
    at com.github.sstone.amqp.Consumer$$anonfun$onChannel$1.apply(Consumer.scala:44) ~[amqp-client_2.10-1.1.jar:na]
    at com.github.sstone.amqp.Consumer$$anonfun$onChannel$1.apply(Consumer.scala:44) ~[amqp-client_2.10-1.1.jar:na]
    at scala.collection.immutable.List.foreach(List.scala:309) ~[scala-library.jar:na]
    at com.github.sstone.amqp.Consumer.onChannel(Consumer.scala:44) ~[amqp-client_2.10-1.1.jar:na]
    at com.github.sstone.amqp.ChannelOwner.setup(ChannelOwner.scala:81) ~[amqp-client_2.10-1.1.jar:na]
    at com.github.sstone.amqp.ChannelOwner$$anonfun$3.applyOrElse(ChannelOwner.scala:86) ~[amqp-client_2.10-1.1.jar:na]
    at com.github.sstone.amqp.ChannelOwner$$anonfun$3.applyOrElse(ChannelOwner.scala:84) ~[amqp-client_2.10-1.1.jar:na]
    at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:33) ~[scala-library.jar:na]
    at akka.actor.FSM$class.processEvent(FSM.scala:576) ~[akka-actor_2.10-2.1.4.jar:na]
    at com.github.sstone.amqp.ChannelOwner.processEvent(ChannelOwner.scala:41) ~[amqp-client_2.10-1.1.jar:na]
    at akka.actor.FSM$class.akka$actor$FSM$$processMsg(FSM.scala:570) ~[akka-actor_2.10-2.1.4.jar:na]
    at akka.actor.FSM$$anonfun$receive$1.applyOrElse(FSM.scala:564) ~[akka-actor_2.10-2.1.4.jar:na]
    at akka.actor.ActorCell.receiveMessage(ActorCell.scala:425) [akka-actor_2.10-2.1.4.jar:na]
    at akka.actor.ActorCell.invoke(ActorCell.scala:386) [akka-actor_2.10-2.1.4.jar:na]
    at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:230) [akka-actor_2.10-2.1.4.jar:na]
    at akka.dispatch.Mailbox.run(Mailbox.scala:212) [akka-actor_2.10-2.1.4.jar:na]
    at akka.dispatch.ForkJoinExecutorConfigurator$MailboxExecutionTask.exec(AbstractDispatcher.scala:506) [akka-actor_2.10-2.1.4.jar:na]
    at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:262) [scala-library.jar:na]
    at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:975) [scala-library.jar:na]
    at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1478) [scala-library.jar:na]
    at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:104) [scala-library.jar:na]
Caused by: com.rabbitmq.client.ShutdownSignalException: channel error; reason: {#method<channel.close>(reply-code=404, reply-text=NOT_FOUND - no exchange 'myexchange' in vhost '/', class-id=40, method-id=10), null, ""}
    at com.rabbitmq.utility.ValueOrException.getValue(ValueOrException.java:67) ~[amqp-client-3.0.1.jar:na]
    at com.rabbitmq.utility.BlockingValueOrException.uninterruptibleGetValue(BlockingValueOrException.java:33) ~[amqp-client-3.0.1.jar:na]
    at com.rabbitmq.client.impl.AMQChannel$BlockingRpcContinuation.getReply(AMQChannel.java:343) ~[amqp-client-3.0.1.jar:na]
    at com.rabbitmq.client.impl.AMQChannel.privateRpc(AMQChannel.java:216) ~[amqp-client-3.0.1.jar:na]
    at com.rabbitmq.client.impl.AMQChannel.exnWrappingRpc(AMQChannel.java:118) ~[amqp-client-3.0.1.jar:na]
    ... 25 common frames omitted
com.rabbitmq.client.ShutdownSignalException: channel error; reason: {#method<channel.close>(reply-code=404, reply-text=NOT_FOUND - no exchange 'myexchange' in vhost '/', class-id=40, method-id=10), null, ""}
    at com.rabbitmq.client.impl.ChannelN.asyncShutdown(ChannelN.java:473) ~[amqp-client-3.0.1.jar:na]
    at com.rabbitmq.client.impl.ChannelN.processAsync(ChannelN.java:313) ~[amqp-client-3.0.1.jar:na]
    at com.rabbitmq.client.impl.AMQChannel.handleCompleteInboundCommand(AMQChannel.java:144) ~[amqp-client-3.0.1.jar:na]
    at com.rabbitmq.client.impl.AMQChannel.handleFrame(AMQChannel.java:91) ~[amqp-client-3.0.1.jar:na]
    at com.rabbitmq.client.impl.AMQConnection$MainLoop.run(AMQConnection.java:533) ~[amqp-client-3.0.1.jar:na]

**Comment by sstone** [comment]
Hi,
You get this error because you are trying to use exchange "myexchange" which does not exist. Is it on purpose (testing...) ?
I think that you are trying to use a consumer for bindings that do not exist, and you get a "declare exchange/crash/start again/declare exchange/crash ...." loop. Can you please describe how you use the client more precisely ?
By the way, you should upgrade to version 1.2 if possible, which has better error handling 
Thanks

**Comment by AsadR** [comment]
Hi,

I'm getting a similar problem running 1.3-ML1 (Scala version 2.10.3). Sometimes, randomly, I get an "exchange not found" error and the binding attempt being made in an infinite loop.

Was this a known issue and fixed in ML2/3 or is it a logical error on my side?

```
Feb 07 12:04:25 ip-10-1-10-64 com.rabbitmq.client.ShutdownSignalException:  channel error; reason: {#method<channel.close>(reply-code=404, reply-text=NOT_FOUND - no exchange '1395' in vhost '/', class-id=50, method-id=20), null, ""}
Feb 07 12:04:29 ip-10-1-10-64 07:  04:44.456 [DEBUG] processing AddBinding(Binding(ExchangeParameters(1395,false,direct,false,true,Map()),QueueParameters(,false,false,true,true,Map()),)) 
Feb 07 12:04:29 ip-10-1-10-64 07:  04:44.476 [WARN ] unhandled event Error(AddBinding(Binding(ExchangeParameters(1395,false,direct,false,true,Map()),QueueParameters(,false,false,true,true,Map()),)),java.io.IOException) in state Connected 
Feb 07 12:04:29 ip-10-1-10-64 com.rabbitmq.client.ShutdownSignalException:  channel error; reason: {#method<channel.close>(reply-code=404, reply-text=NOT_FOUND - no exchange '1395' in vhost '/', class-id=50, method-id=20), null, ""}
Feb 07 12:04:29 ip-10-1-10-64 07:  04:44.539 [DEBUG] processing AddBinding(Binding(ExchangeParameters(1395,false,direct,false,true,Map()),QueueParameters(,false,false,true,true,Map()),)) 
Feb 07 12:04:29 ip-10-1-10-64 com.rabbitmq.client.ShutdownSignalException:  channel error; reason: {#method<channel.close>(reply-code=404, reply-text=NOT_FOUND - no exchange '1395' in vhost '/', class-id=50, method-id=20), null, ""}
Feb 07 12:04:29 ip-10-1-10-64 07:  04:44.551 [WARN ] unhandled event Error(AddBinding(Binding(ExchangeParameters(1395,false,direct,false,true,Map()),QueueParameters(,false,false,true,true,Map()),)),java.io.IOException) in state Disconnected 
Feb 07 12:04:29 ip-10-1-10-64 07:  04:44.654 [DEBUG] processing AddBinding(Binding(ExchangeParameters(1395,false,direct,false,true,Map()),QueueParameters(,false,false,true,true,Map()),)) 
Feb 07 12:04:29 ip-10-1-10-64 07:  04:44.661 [DEBUG] processing AddBinding(Binding(ExchangeParameters(1395,false,direct,false,true,Map()),QueueParameters(,false,false,true,true,Map()),)) 
...
```

**Comment by sstone** [comment]
Hi,
It looks like a real bug, I'll have a fix ready this afternoon. The problem is that the exchange specified in the binding is not created when its 'passive' parameter is set to false, as it should be.
 I'll have a fix ready this afternoon, but it might take a few hours before it is uploaded to maven central.
There is a workaround: it you just use a fixed set of exchanges, you just need to create them (with the admin tool, the web gui, ...) before you start your applications.  
Thanks!

**Comment by sstone** [comment]
See issue #33. Can you please try the latest 1.3-SNAPSHOT from the sonatype repo (https://oss.sonatype.org/content/repositories/snapshots/) ? It should fix the problem.

**Comment by burgup** [comment]
It seems the looping issue is still not fixed. I tried with 1.3-ML5 and 1.3-SNAPSHOT, I still see the following message repeatedly when the exchange is not present and it ends up creating lots of anonymous queues within a short span.

[akka://actorsystem/user/amqp-consumer/newConn/$a] shutdown
com.rabbitmq.client.ShutdownSignalException: channel error; reason: {#method<channel.close>(reply-code=404, reply-text=NOT_FOUND - no exchange 'TestExchange' in vhost '/', class-id=40, method-id=10), null, ""}
 at com.rabbitmq.client.impl.ChannelN.asyncShutdown(ChannelN.java:474)
 at com.rabbitmq.client.impl.ChannelN.processAsync(ChannelN.java:315)
 at com.rabbitmq.client.impl.AMQChannel.handleCompleteInboundCommand(AMQChannel.java:144)
 at com.rabbitmq.client.impl.AMQChannel.handleFrame(AMQChannel.java:91)
 at com.rabbitmq.client.impl.AMQConnection$MainLoop.run(AMQConnection.java:551)

**Comment by sstone** [comment]
Hi Burgup,
Can you please explain what you're doing more precisely ? (ideally a description in plain English and a few lines of code with the actual queue/exchange parameters). By anonymous queue I guess that you mean queues with a random, broker-generated name ?
Thanks

**Comment by burgup** [comment]
> By anonymous queue I guess that you mean queues with a random, broker-generated name ?
> That's right.

Here is the sample code snippet that reproduces the problem. This code tries to add a binding to a non existing exchange. If exchange is already present, then things work correctly.

```
 final val timeout = 2 seconds
val exchange = ExchangeParameters(name = "Non_existing_Exchange", exchangeType = "topic", durable = true, passive = true)
val messageTTL = new Integer(3600000)
val queue = QueueParameters(name = "", passive = false, exclusive = true, autodelete = true, args = Map("x-message-ttl" -> messageTTL))

ConnectionOwner.createChildActor(conn, Consumer.props(
  listener = Some(self),
  init = List(AddBinding(Binding(exchange, queue, "#")))
), name = None, timeout)
```

**Comment by sstone** [comment]
Hi,
In your exchange declaration the 'passive' parameter is set to 'true': the library will just check that the exchange exist but will not create it if it doesn't. You probably meant to use 'passive=false' instead 
Thanks

**Comment by burgup** [comment]
I'm using "passive=true" intentionally.  The above code snippet is just to reproduce the problem. My application is reading the exchange name from config, and this problem happens when the configuration is incorrect.

**Comment by sstone** [comment]
Ok, so there's no problem and I can keep this issue closed ?
Thanks

**Comment by burgup** [comment]
No. There is a problem. Irrespective of the config, there should not be infinite loop creating multiple queues when the exchange is not present. If the exchange is not present, the binding should just fail with error. Please try with the above code snippet, and you can see the problem first hand.

**Comment by sstone** [comment]
To avoid this loop you could create an unbound consumer:
val consumer = ConnectionOwner.createChildActor(conn, Consumer.props(listener = Some(self))

And then send it a binding message:
consumer ! AddBinding(Binding(exchangeParams, queueParams, "my_key"))

This will give you better error feedback and the consumer will no longer enter a crash/restart/crash loop.

**Comment by zhangj311** [comment]
Hi, I am running into the same problem with versoin 3.5.3 and spring-rabbit-1.4.5,can you give same suggestions?
